File: JavaScript/13-Number_Line_Jumps.js
/**
 * Created on Wed Feb 21 2023
 * 
 * @author Carlos Páez
 */
'use strict';

const fs = require('fs')

process.stdin.resume()
process.stdin.setEncoding('utf-8')

let inputString = ''
let currentLine = 0

/* Reading the input from the console. */
process.stdin.on('data', function (inputStdin) {
    inputString += inputStdin
})

/* Reading the input from the console. */
process.stdin.on('end', function () {
    inputString = inputString.split('n')
    main()
})


/**
 * It reads the input string and returns the current line.
 * @returns The current line of the inputString.
 */
function readLine () {
    return inputString[ currentLine++ ]
}


/*
 * Complete the `kangaroo` function below.
 *
 * The function is expected to return a STRING.
 * The function accepts following parameters:
 *  1. INTEGER x1
 *  2. INTEGER v1
 *  3. INTEGER x2
 *  4. INTEGER v2
 */

/**
 * If the kangaroos are not moving at the same speed, and the distance between them is divisible by the
 * difference in their speeds, then they will meet
 * @param x1 - starting position for kangaroo 1
 * @param v1 - Kangaroo 1's jump distance (in meters)
 * @param x2 - starting position of kangaroo 2
 * @param v2 - the speed of the second kangaroo
 * @returns a string.
 */
function kangaroo (x1, v1, x2, v2) {
    const maxLimit = 10000

    const invalidCondition = (
        x1 < 0 || x1 > maxLimit || x2 < 0 || x2 > maxLimit ||
        v1 < 0 || v1 > maxLimit || v2 < 0 || v2 > maxLimit ||
        x1 > x2 || v1 < v2
    )

    if (invalidCondition) return 'NO'

    if (v1 !== v2 && (x2 - x1) % (v2 - v1) === 0) return 'YES'

    return 'NO'
}

function main () {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const firstMultipleInput = readLine().replace(/\s+$/g, '').split(' ');

    const x1 = parseInt(firstMultipleInput[ 0 ], 10);

    const v1 = parseInt(firstMultipleInput[ 1 ], 10);

    const x2 = parseInt(firstMultipleInput[ 2 ], 10);

    const v2 = parseInt(firstMultipleInput[ 3 ], 10);

    const result = kangaroo(x1, v1, x2, v2);

    ws.write(result + '\n');

    ws.end();
}

File: Python/13-Number_Line_Jumps.py
# -*- coding: utf-8 -*-
""" 
Created on Tue Feb 21 2023

@author Carlos Páez
"""

#
# Complete the 'kangaroo' function below.
#
# The function is expected to return a STRING.
# The function accepts following parameters:
#  1. INTEGER x1
#  2. INTEGER v1
#  3. INTEGER x2
#  4. INTEGER v2
#


def kangaroo(x1, v1, x2, v2) -> str:
    """
    If the kangaroos are not moving at the same speed, and the distance between them is a multiple
    of the difference in their speeds, then they will meet. 

    :param x1: starting position for kangaroo 1
    :param v1: Kangaroo 1's jump distance
    :param x2: starting position of kangaroo 2
    :param v2: the speed of the second kangaroo
    :return: a string.
    """
    max_limit = 10000
    if not (0 <= x1 < x2 <= max_limit) or not (1 <= v1 <= max_limit) or not (1 <= v2 <= max_limit):
        return 'NO'
    if x1 < x2 and v1 < v2:
        return 'NO'
    else:
        return 'YES' if v1 != v2 and (x2 - x1) % (v2 - v1) == 0 else 'NO'


if __name__ == '__main__':
    fptr = open('./output.txt', 'w')

    first_multiple_input = input().rstrip().split()

    x1 = int(first_multiple_input[0])
    v1 = int(first_multiple_input[1])
    x2 = int(first_multiple_input[2])
    v2 = int(first_multiple_input[3])

    result = kangaroo(x1, v1, x2, v2)

    fptr.write(result + '\n')

    fptr.close()
